// @opensource-technologies/typescript-data-structure-library/src/LinkedList.ts
export class ListNode<T> {
  public value: T;
  public next: ListNode<T> | null = null;

  constructor(value: T) {
    this.value = value;
  }
}

export class LinkedList<T> {
  private head: ListNode<T> | null = null;

  // Append a node to the end of the list
  public append(value: T): void {
    const newNode = new ListNode(value);

    if (this.head === null) {
      this.head = newNode;
      return;
    }

    let current: ListNode<T> = this.head;
    while (current.next !== null) {
      current = current.next;
    }

    current.next = newNode;
  }

  // Prepend a node to the start of the list
  public prepend(value: T): void {
    const newNode = new ListNode(value);
    newNode.next = this.head;
    this.head = newNode;
  }

  // Remove a node by value
  public remove(value: T): void {
    if (this.head === null) return;

    if (this.head.value === value) {
      this.head = this.head.next;
      return;
    }

    let current: ListNode<T> = this.head;
    while (current.next !== null && current.next.value !== value) {
      current = current.next;
    }

    if (current.next !== null) {
      current.next = current.next.next;
    }
  }

  // Find a node by value
  public find(value: T): ListNode<T> | null {
    let current = this.head;
    while (current !== null && current.value !== value) {
      current = current.next;
    }
    return current;
  }

  // Convert the list to an array
  public toArray(): T[] {
    const elements: T[] = [];
    let current = this.head;

    while (current !== null) {
      elements.push(current.value);
      current = current.next;
    }

    return elements;
  }

  // Insert a node at a specific index
  public insertAt(value: T, index: number): void {
    if (index < 0) throw new Error("Index out of bounds");

    const newNode = new ListNode(value);

    if (index === 0) {
      newNode.next = this.head;
      this.head = newNode;
      return;
    }

    let current = this.head;
    let currentIndex = 0;

    while (current !== null && currentIndex < index - 1) {
      current = current.next;
      currentIndex++;
    }

    if (current === null) throw new Error("Index out of bounds");

    newNode.next = current.next;
    current.next = newNode;
  }

  // Get the length of the list
  public length(): number {
    let count = 0;
    let current = this.head;

    while (current !== null) {
      count++;
      current = current.next;
    }

    return count;
  }

  // Reverse the list
  public reverse(): void {
    let prev: ListNode<T> | null = null;
    let current = this.head;
    let next: ListNode<T> | null = null;

    while (current !== null) {
      next = current.next;
      current.next = prev;
      prev = current;
      current = next;
    }

    this.head = prev;
  }

  // Check if the list is empty
  public isEmpty(): boolean {
    return this.head === null;
  }

  // Clear the list
  public clear(): void {
    this.head = null;
  }
}
