// @opensource-technologies/typescript-data-structure-library/dist/LinkedList.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LinkedList = exports.ListNode = void 0;
class ListNode {
    constructor(value) {
        this.next = null;
        this.value = value;
    }
}
exports.ListNode = ListNode;
class LinkedList {
    constructor() {
        this.head = null;
    }
    // Append a node to the end of the list
    append(value) {
        const newNode = new ListNode(value);
        if (this.head === null) {
            this.head = newNode;
            return;
        }
        let current = this.head;
        while (current.next !== null) {
            current = current.next;
        }
        current.next = newNode;
    }
    // Prepend a node to the start of the list
    prepend(value) {
        const newNode = new ListNode(value);
        newNode.next = this.head;
        this.head = newNode;
    }
    // Remove a node by value
    remove(value) {
        if (this.head === null)
            return;
        if (this.head.value === value) {
            this.head = this.head.next;
            return;
        }
        let current = this.head;
        while (current.next !== null && current.next.value !== value) {
            current = current.next;
        }
        if (current.next !== null) {
            current.next = current.next.next;
        }
    }
    // Find a node by value
    find(value) {
        let current = this.head;
        while (current !== null && current.value !== value) {
            current = current.next;
        }
        return current;
    }
    // Convert the list to an array
    toArray() {
        const elements = [];
        let current = this.head;
        while (current !== null) {
            elements.push(current.value);
            current = current.next;
        }
        return elements;
    }
    // Insert a node at a specific index
    insertAt(value, index) {
        if (index < 0)
            throw new Error("Index out of bounds");
        const newNode = new ListNode(value);
        if (index === 0) {
            newNode.next = this.head;
            this.head = newNode;
            return;
        }
        let current = this.head;
        let currentIndex = 0;
        while (current !== null && currentIndex < index - 1) {
            current = current.next;
            currentIndex++;
        }
        if (current === null)
            throw new Error("Index out of bounds");
        newNode.next = current.next;
        current.next = newNode;
    }
    // Get the length of the list
    length() {
        let count = 0;
        let current = this.head;
        while (current !== null) {
            count++;
            current = current.next;
        }
        return count;
    }
    // Reverse the list
    reverse() {
        let prev = null;
        let current = this.head;
        let next = null;
        while (current !== null) {
            next = current.next;
            current.next = prev;
            prev = current;
            current = next;
        }
        this.head = prev;
    }
    // Check if the list is empty
    isEmpty() {
        return this.head === null;
    }
    // Clear the list
    clear() {
        this.head = null;
    }
}
exports.LinkedList = LinkedList;
